Refactor parsing of .ocd parameter strings

dl3sdo opened the pull request.

dg0yt commented:
Adding an abstraction might be helpful, but IMO it should take a different approach. The reason why it wasn't implemented like this is efficiency.
- Memory allocation (on the heap) is always expensive. 
  - The `OcdParameterString` needs to allocate memory for the vectors. Here, we also don't know how in advance how much memory to allocate. So we may even see reallocations.
  - The individual strings need memory allocations, too.
- In the end, the actual data is scattered somewhere in memory. Accessing scattered data is more expensive than accessing data which is contiguous and already in cache.

I admit it is not a hot path, even for big OCD files. But allocating scattered memory for single-letter keys is something I would prefer to avoid.

I would suggest an approach which operates directly on the original data. For example, it could be an "OCD parameter stream reader", constructed with the input string, providing `bool atEnd() const`, `bool readNext()`, `QStringRef key() const`, `QStringRef value() const`. Somewhat similar to `QXmlStreamReader`. The reader would encapsulate handling the current reading position, but it wouldn't need any heap memory.

dl3sdo commented:
@dg0yt: You are right that it makes no sense to keep everything in memory if only sequential processing of the parameters is applied as it is currently done (although the runtime overhead is of course negligible).
I adopted your proposal for the class functions but prefer to have key() return plain char. I implemented atEnd() but only use it class internally and in one test case.
I implemented the test cases as a system test although the scope it that of a unit test (however I ran into various compiler and linker issues when trying to implement as unit test).

dg0yt commented on src/fileformats/ocd_file_import.h:
```diff
@@ -118,6 +118,21 @@ class OcdFileImport : public Importer
+		
+	private:
+		const QString& param_string;
+		int param_ptr;
```
I would prefer `current_pos` or just `pos`, because it is not a C pointer.

dg0yt commented on src/fileformats/ocd_file_import.h:
```diff
@@ -362,6 +377,11 @@ class OcdFileImport : public Importer
 	int ocd_version;
 };
 
+inline
```
Just move it into the cpp file. It is used only there, and it will still be inlined based on compiler decisions.

(Extended comment: The whole class declaration could be moved to the cpp file, but this would prevent unit testing. The whole class could be a separate translation unit, and then the `atEnd` definition would indeed need go to the header to allow inlining. Today's preferred form is to put the body into the class definition instead of an explicit `inline` - again in order to let the compile decide about inlining.)

dg0yt reviewed (changes requested):
This is an excellent contribution now. Mni tnx! :rocket:

dl3sdo commented on src/fileformats/ocd_file_import.cpp:
```diff
@@ -2449,4 +2412,57 @@ void OcdFileImport::handleStrings(const OcdFile<F>& file, std::initializer_list<
+	if (param_ptr)
+		return param_string.at(param_ptr).toLatin1();
+	else
+		return '\f';	// return generic '\f' for 'first' parameter
```
The idea behind the \f is that processing of the 'first' parameter could also be done inside the loop:
```
const QString name = parameters.value().toString();		// 'first' parameter
while (parameters.readNext())
{
switch (parameters.key())
		{
		case 'n':
			number = param_value.toInt(&number_ok);
			break;
```
could be changed to
```
do
{
switch (parameters.key())
		{
		case '\f':
			name = parameters.value().toString();	// 'first' parameter
			break;
		case 'n':
			number = param_value.toInt(&number_ok);
			break;
.....
} while (parameters.readNext());
```
Since \f is not used as a regular key (but f is) I decided to chose \f in allusion to 'first'.
If the parameter string would be empty, the generic '\0' would fall through and the while below end any futher parsing.

dg0yt replied:
Okay, `'\f'  != 'f'`.

If we want a cute name, we should use a `constexpr` member function returning the desired character. And `\0` would still work. Reading at the end is an invalid operation anyways, so there is no conflict.

dl3sdo replied:
I think the implementation of '\f' is acceptable since the scope of that helpe class is very limited and since '\f' is actually not used by the parent class when processing parameter strings. Besides that, the arbitrarily chosen key replacement is well documented.
I don't know how to implement such a constexpr member function, so please make a proposal.

dl3sdo commented:
@dg0yt: thank you!

dl3sdo commented:
I agree with the additional changes, everything runs fine as expected.
After being merged into master I would use it for continuing the work on #959.

dg0yt commented:
Tests and public API are fine, so let's merge as is now.
